Simplenote/ExcerptMaker.swift

```diff
@@ -0,0 +1,61 @@
+import Foundation
+
+// MARK: - ExcerptMaker: Generate excerpt from a note with specified keywords
+//
+@objc
+final class ExcerptMaker: NSObject {
+    private var regexp: NSRegularExpression?
+    private var keywords: [String]?
+
+    /// Update with search keywords
+    ///
+    @objc
+    func update(withKeywords keywords: [String]?) {
+        guard keywords != self.keywords else {
+            return
+        }
+
+        self.keywords = keywords
+
+        guard let keywords = keywords, !keywords.isEmpty else {
+            regexp = nil
+            return
+        }
+
+        regexp = NSRegularExpression.regexForExcerpt(withKeywords: keywords)
+    }
+
+    /// Generate and return excerpt from the note. Excerpt is based on keywords
+    ///
+    func excerpt(from note: Note) -> String? {
+        guard let regexp = regexp, let content = note.content else {
+            return note.bodyPreview
+        }
+
+        let foldedContent = content.folding(options: [.diacriticInsensitive], locale: .current)
+
+        let bodyRange = NoteContentHelper.structure(of: foldedContent).body
+        guard bodyRange.location != NSNotFound else {
+            return nil
+        }
+
+        let excerptRange = regexp.rangeOfFirstMatch(in: foldedContent,
+                                                    options: [],
+                                                    range: bodyRange)
+        guard excerptRange.location != NSNotFound else {
+            return note.bodyPreview
+        }
+
+        var excerpt: String = {
+            if foldedContent.nsString.length == content.nsString.length {
+                return content.nsString.substring(with: excerptRange)
+            }
+            return foldedContent.nsString.substring(with: excerptRange)
+        }()
+        excerpt = excerpt.nsString.replacingNewlinesWithSpaces()
+        if excerptRange.location > bodyRange.location {
+            excerpt = "…" + excerpt
+        }
+        return excerpt
+    }
```

Comment:
The resulting string will not contain diacritics (unless I'm missing something). Is that expected?

Reply:
It will return substring from the original string (containing diacritics), unless folded and original strings have different lengths.
I believe that in our case folding will produce the string of the same length as the original one, but just in case it won't I added a check.

It's a pity that NSRegularExpression doesn't support diacritic insensitive matching and the approach is not ideal.
Another options that we have:
1) include all possible diactric symbols when creating a pattern from keywords. For example: `test` will become `t[eèéêëēėę]st`
2) Don't use regexp. Split string by words and compare keywords to each word (using case and diactric insensitive mode). Stop when we have the first match. Get some words around it.

What do you think?